src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.IO;
+using System.Security.Cryptography.X509Certificates;
+using System.Security.Cryptography;
+
+namespace System.Net.Http
+{
+    internal partial class CurlHandler : HttpMessageHandler
+    {
+
+        private sealed class ClientCertProvider : IDisposable
+        {
+            internal readonly string _certificateFile;
+            internal readonly string _password;
+
+            private ClientCertProvider(X509Certificate2Collection certs)
+            {
+                Debug.Assert(certs.Count >0);
+                try
+                {
+                    _password = GeneratePassword();
+                    byte[] pfxData = certs.Export(X509ContentType.Pkcs12, _password);
+                    _certificateFile = Path.GetTempFileName();
+                    File.WriteAllBytes(_certificateFile, pfxData);
+                }
+                catch (Exception e)
+                {
+                    throw new HttpRequestException(SR.net_http_client_execution_error, e);
+                }
+            }
+
+            public void Dispose()
+            {
+                if (_certificateFile == null)
+                {
+                    return;
+                }
+                try
+                {
+                    File.Delete(_certificateFile);
+                }
+                catch (IOException e)
+                {
+                    VerboseTrace("Deleting temporary file:" + e);
+                }
+            }
+
+            internal static ClientCertProvider Create(ClientCertificateOption option, string uriScheme)
+            {
+                if (option == ClientCertificateOption.Manual || !string.Equals(uriScheme, UriSchemeHttps))
+                {
+                    return null;
+                }
+
+                X509Certificate2Collection clientCerts = new X509Certificate2Collection();
+                using (var myStore = new X509Store())
+                {
+                    myStore.Open(OpenFlags.ReadOnly);
+                    X509Certificate2Collection certificates = myStore.Certificates;
+
+                    certificates = certificates.Find(X509FindType.FindByKeyUsage, X509KeyUsageFlags.DigitalSignature, true);
```

Comment:
There are times when one looks at code and is compelled to ask "did you test this?".  In this case, I'm pretty sure the answer has to be "no", because I haven't written `Find` on Unix yet:  It always emits the empty set.

https://github.com/dotnet/corefx/blob/master/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509StoreProvider.cs#L24

This flow would be incredibly expensive, anyways.  It would create two additional collections and the found certs in each collection would be cloned for each call to Find.

This should really just be one for/foreach doing all three tests for performance reasons.

Though remember that for both KeyUsage and ExtendedKeyUsage, if the extension is not present on the certificate it is considered valid for all purposes.

Aside: For some reason I thought that the Find method was mostly implemented in Windows code I was going to have to duplicate, but it appears to be mostly managed... it seems me finishing that for Unix would be somewhat helpful. (Though I do still feel that chained calls to it like you have here have too much GC+Finalizer overhead)